--- cisco-vpnclient-4.8.01.0640_k9/vpnclient/interceptor.c	2008-07-01 04:19:42.207419395 +0200
+++ cisco-vpnclient-4.8.01.0640_k9-shadzik/vpnclient/interceptor.c	2008-07-01 04:18:09.427011880 +0200
@@ -24,6 +24,7 @@
 #include <linux/in.h>
 #include <linux/ppp_defs.h>
 #include <net/ip.h>
+#include <net/net_namespace.h>
 #include <linux/ip.h>
 #include <linux/udp.h>
 #include <net/protocol.h>
@@ -306,6 +307,7 @@
     static struct packet_type dummy_pt;
 
     struct net_device *dp = NULL;
+    struct net *net;
     struct packet_type *default_pt = NULL;
     int error = VPNIFUP_SUCCESS, num_target_devices;
 
@@ -361,11 +363,12 @@
     kernel_memset(&Bindings, 0, sizeof(Bindings));
 
     dp = NULL;
+    net = dp->nd_net;
     num_target_devices = 0;
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,22)
-    for_each_netdev(dp)
+    for_each_netdev(net, dp)
 #else
-    for (dp = dev_base; dp != NULL; dp = dp->next)
+    for (dp = dev_base; dp != NULL; dp = dp->dev_list.next)
 #endif
     {
         if (add_netdev(dp) == 0)
--- cisco-vpnclient-4.8.01.0640_k9/vpnclient/Makefile	2007-08-22 21:30:31.000000000 +0200
+++ cisco-vpnclient-4.8.01.0640_k9-shadzik/vpnclient/Makefile	2008-07-01 03:37:56.054335689 +0200
@@ -12,7 +12,7 @@
 SOURCE_OBJS := linuxcniapi.o frag.o IPSecDrvOS_linux.o interceptor.o linuxkernelapi.o
 
 ifeq ($(SUBARCH),x86_64)
-CFLAGS += -mcmodel=kernel -mno-red-zone
+EXTRA_CFLAGS += -mcmodel=kernel -mno-red-zone
 NO_SOURCE_OBJS := libdriver64.so
 else
 NO_SOURCE_OBJS := libdriver.so
--- cisco-vpnclient-4.8.01.0640_k9/vpnclient/GenDefs.h	2007-08-22 21:30:31.000000000 +0200
+++ cisco-vpnclient-4.8.01.0640_k9-shadzik/vpnclient/GenDefs.h	2008-07-01 03:40:19.151196714 +0200
@@ -105,6 +105,7 @@
 #define _INTPTR_T_DEFINED
 #endif
 
+#define _UINTPTR_T_DEFINED
 #ifndef _UINTPTR_T_DEFINED
 #if defined(_LP64)
 #warning 64 bit
